src/ESLint.Formatter/sarif-with-rules.js

```diff
@@ -0,0 +1,145 @@
+/**
+ * @fileoverview SARIF v2.0 formatter
+ * @author Chris Meyer
+ */
+
+"use strict";
+
+//------------------------------------------------------------------------------
+// Helper Functions
+//------------------------------------------------------------------------------
+
+/**
+ * Returns the severity of warning or error
+ * @param {Object} message message object to examine
+ * @returns {string} severity level
+ * @private
+ */
+function getResultLevel(message) {
+    if (message.fatal || message.severity === 2) {
+        return "error";
+    }
+    return "warning";
+}
+
+
+//------------------------------------------------------------------------------
+// Public Interface
+//------------------------------------------------------------------------------
+
+module.exports = function(results, rules) {
+    const sarifLog = {
+        version: "2.0.0",
+        $schema: "http://json.schemastore.org/sarif-2.0.0",
+        runs: [
+            {
+                tool: {
+                    name: "ESLint"
+                }
+            }
+        ]
+    };
+
+    const sarifFiles = {};
+    const sarifRules = {};
+    const sarifResults = [];
+
+    results.forEach(result => {
+
+        if (typeof sarifFiles[result.filePath] === "undefined") {
+
+            // Create a new entry in the files dictionary.
+            sarifFiles[result.filePath] = {
```

Comment:
>sarifFiles [](start = 12, length = 10)

are we going to populate any other file data? if it's just the URI, not entirely sure whether persisting it is helpful. It's not a bad thing, per se, but will consume log file size. mostly these entries exist to allow for annotation with other data, hashes, mime type,  size of file, etc.

Reply:
I am able to read the file contents so I've added that code.

---
In reply to: [249141534](https://github.com/Microsoft/sarif-sdk/pull/1222#discussion_r249141534) [](ancestors = 249141534,249139155)

Reply:
I also added a mechanism to control embedding: set SARIF_ESLINT_EMBED=true

---
In reply to: [249579009](https://github.com/Microsoft/sarif-sdk/pull/1222#discussion_r249579009) [](ancestors = 249579009,249141534,249139155)